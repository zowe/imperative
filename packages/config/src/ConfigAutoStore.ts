/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import * as lodash from "lodash";
import { ICommandArguments, IHandlerParameters } from "../../cmd";
import { ICommandHandlerRequire } from "../../cmd/src/doc/handler/ICommandHandlerRequire";
import { ICommandProfileAuthConfig } from "../../cmd/src/doc/profiles/definition/ICommandProfileAuthConfig";
import { IConfigLoadedProfile } from "./doc/IConfigLoadedProfile";
import * as ConfigUtils from "./ConfigUtils";
import { AbstractAuthHandler } from "../../imperative/src/auth/handlers/AbstractAuthHandler";
import { ImperativeConfig } from "../../utilities";
import { ISession } from "../../rest/src/session/doc/ISession";
import { Session } from "../../rest/src/session/Session";
import { AUTH_TYPE_TOKEN } from "../../rest/src/session/SessConstants";
import { Logger } from "../../logger";

/**
 * Class to manage automatic storage of properties in team config.
 */
export class ConfigAutoStore {
    /**
     * Finds the token auth handler class for a team config profile.
     * @param profilePath JSON path of profile
     * @param cmdArguments CLI arguments which may specify a profile
     * @returns Auth handler class or undefined if none was found
     */
    public static findAuthHandlerForProfile(profilePath: string, cmdArguments: ICommandArguments): AbstractAuthHandler | undefined {
        const config = ImperativeConfig.instance.config;
        const profileType = lodash.get(config.properties, `${profilePath}.type`);
        const profile = config.api.profiles.get(profilePath.replace(/profiles\./g, ""));

        if (profile == null || profileType == null) {  // Profile must exist and have type defined
            return;
        } else if (profileType === "base") {
            if (profile.tokenType == null) {  // Base profile must have tokenType defined
                return;
            }
        } else {
            if (profile.basePath == null) {  // Service profiles must have basePath defined
                return;
            } else if (profile.tokenType == null) {  // If tokenType undefined in service profile, fall back to base profile
                const baseProfileName = ConfigUtils.getActiveProfileName("base", cmdArguments);
                return this.findAuthHandlerForProfile(config.api.profiles.expandPath(baseProfileName), cmdArguments);
            }
        }

        const authConfigs: ICommandProfileAuthConfig[] = [];
        ImperativeConfig.instance.loadedConfig.profiles.forEach((profCfg) => {
            if ((profCfg.type === profileType || profCfg.type === "base") && profCfg.authConfig != null) {
                authConfigs.push(...profCfg.authConfig);
            }
        });

        for (const authConfig of authConfigs) {
            const authHandler: ICommandHandlerRequire = require(authConfig.handler);
            const authHandlerClass = new authHandler.default();

            if (authHandlerClass instanceof AbstractAuthHandler) {
                const promptParams = authHandlerClass.getPromptParams()[0];

                if (profile.tokenType === promptParams.defaultTokenType) {
                    return authHandlerClass;  // Auth service must have matching token type
                }
            }
        }
    }

    /**
     * Finds the highest priority layer where a profile is stored.
     * @param loadedProfile
     * @returns User and global properties
     */
    public static getPriorityLayer(loadedProfile: IConfigLoadedProfile): { user: boolean, global: boolean } {
        return {
            user: Object.values(loadedProfile.properties).every(v => v.user),
            global: Object.values(loadedProfile.properties).some(v => v.global)
        };
    }

    /**
     * Stores session config properties into a team config profile.
     * @param params CLI handler parameters object
     * @param sessCfg Session config containing properties to store
     * @param propsToStore Names of properties that should be stored
     */
    public static async storeSessCfgProps(params: IHandlerParameters, sessCfg: { [key: string]: any }, propsToStore: string[]): Promise<void> {
        const config = ImperativeConfig.instance.config;
        // TODO Which autoStore value should take priority if it conflicts between layers
        if (propsToStore.length == 0 || !config?.exists || !config.properties.autoStore) {
            return;
        }

        let profileProps = propsToStore.map(propName => propName === "hostname" ? "host" : propName);
        const profileData = this.findActiveProfile(params, profileProps);
        if (profileData == null) {
            return;
        }
        const [profileType, profileName] = profileData;
        const profilePath = config.api.profiles.expandPath(profileName);

        // Replace user and password with tokenValue if tokenType is defined in config
        if (profileProps.includes("user") && profileProps.includes("password") &&
            await this.fetchTokenForSessCfg(params, sessCfg, profilePath)) {
            profileProps = profileProps.filter(propName => propName !== "user" && propName !== "password");
            profileProps.push("tokenValue");
        }

        const beforeLayer = config.api.layers.get();
        const loadedProfile = config.api.profiles.load(profileName);
        const { user, global } = this.getPriorityLayer(loadedProfile);
        config.api.layers.activate(user, global);

        const baseProfileName = ConfigUtils.getActiveProfileName("base", params.arguments);
        const baseProfileObj = lodash.get(config.properties, config.api.profiles.expandPath(baseProfileName));
        const baseProfileSchema = ImperativeConfig.instance.loadedConfig.baseProfile.schema;
        const profileSchema = ImperativeConfig.instance.loadedConfig.profiles.find(p => p.type === profileType).schema;

        for (const propName of profileProps) {
            let propProfilePath = profilePath;
            // Determine if property should be stored in base profile instead
            if (loadedProfile.properties[propName] == null && !loadedProfile.secure?.includes(propName) &&
                (baseProfileObj.properties[propName] != null || baseProfileObj.secure?.includes(propName) ||
                (propName === "tokenValue" && baseProfileObj.properties.tokenType != null))) {
                propProfilePath = config.api.profiles.expandPath(baseProfileName);
            }
            const sessCfgPropName = propName === "host" ? "hostname" : propName;
            const propDefinition = profileSchema.properties[propName] || baseProfileSchema.properties[propName];
            config.set(`${propProfilePath}.properties.${propName}`, sessCfg[sessCfgPropName], {
                secure: propDefinition.secure
            });
        }

        await config.save(false);
        params.response.console.log(`Stored properties in ${config.layerActive().path}: ${profileProps.join(", ")}`);
        // Restore original active layer
        config.api.layers.activate(beforeLayer.user, beforeLayer.global);
    }

    /**
     * Finds the profile where auto-store properties should be saved.
     * @param params CLI handler parameters object
     * @param profileProps List of properties required in the profile schema
     * @returns Tuple containing profile type and name, or undefined if no
     *          profile was found
     */
    private static findActiveProfile(params: IHandlerParameters, profileProps: string[]): [string, string] | undefined {
        const profileTypes = [
            ...(params.definition.profile?.required || []),
            ...(params.definition.profile?.optional || [])
        ];

        for (const profType of profileTypes) {
            const profileMatch = ImperativeConfig.instance.loadedConfig.profiles.find(p => p.type === profType);
            if (profileMatch != null && profileProps.every(propName => propName in profileMatch.schema.properties)) {
                return [profType, ConfigUtils.getActiveProfileName(profType, params.arguments)];
            }
        }
    }

    /**
     * Retrieves token value that will be auto-stored into session config.
     * @param params CLI handler parameters object
     * @param sessCfg Session config with credentials for basic or cert auth
     * @param profilePath JSON path of profile containing tokenType
     * @returns True if auth handler was found and token was fetched
     */
    private static async fetchTokenForSessCfg(params: IHandlerParameters, sessCfg: { [key: string]: any }, profilePath: string): Promise<boolean> {
        const authHandlerClass = this.findAuthHandlerForProfile(profilePath, params.arguments);

        if (authHandlerClass == null) {
            return false;
        }

        const [promptParams, loginHandler] = authHandlerClass.getPromptParams();
        sessCfg.type = AUTH_TYPE_TOKEN;
        sessCfg.tokenType = promptParams.defaultTokenType;
        const baseSessCfg: ISession = { type: sessCfg.type };

        for (const propName of Object.keys(ImperativeConfig.instance.loadedConfig.baseProfile.schema.properties)) {
            const sessCfgPropName = propName === "host" ? "hostname" : propName;
            if (sessCfg[sessCfgPropName] != null) {
                (baseSessCfg as any)[sessCfgPropName] = sessCfg[sessCfgPropName];
            }
        }

        Logger.getAppLogger().info(`Fetching ${sessCfg.tokenType} for ${profilePath}`);
        sessCfg.tokenValue = await loginHandler(new Session(baseSessCfg));
        return true;
    }
}
