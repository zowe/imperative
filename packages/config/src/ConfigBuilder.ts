/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import * as lodash from "lodash";
import { IImperativeConfig } from "../../imperative";
import { Config } from "./Config";
import { IConfig } from "./doc/IConfig";
import { IConfigBuilderOpts } from "./doc/IConfigBuilderOpts";
import { IConfigProfile } from "./doc/IConfigProfile";

export class ConfigBuilder {
    /**
     * Build a new Config object from an Imperative CLI app configuration.
     * @param impConfig The Imperative CLI app configuration.
     * @param opts Options to control aspects of the builder.
     */
    public static async build(impConfig: IImperativeConfig, opts?: IConfigBuilderOpts): Promise<IConfig> {
        opts = opts || {};
        const config: IConfig = Config.empty();

        for (const profile of impConfig.profiles) {
            const properties: { [key: string]: any } = {};
            const secureProps: string[] = [];
            for (const [k, v] of Object.entries(profile.schema.properties)) {
                if (opts.populateProperties && v.includeInTemplate) {
                    if (v.secure) {
                        secureProps.push(k);
                        const propValue = opts.getSecureValue ? await opts.getSecureValue(k, v) : null;
                        if (propValue != null) {
                            // Save this value to be stored securely after profile is built
                            properties[k] = propValue;
                        }
                    } else {
                        if ((v as any).optionDefinition != null) {
                            // Use default value of ICommandOptionDefinition if present
                            properties[k] = (v as any).optionDefinition.defaultValue;
                        }
                        if (properties[k] === undefined) {
                            // Fall back to an empty value
                            properties[k] = this.getDefaultValue(v.type);
                        }
                    }
                }
            }

            // Add the profile to config and set it as default
            lodash.set(config, `profiles.${profile.type}`, {
                type: profile.type,
                properties,
                secure: secureProps
            });

            if (opts.populateProperties) {
                config.defaults[profile.type] = profile.type;
            }
        }

        // Hoist duplicate default properties
        if (config.profiles != null && impConfig.baseProfile != null) {
            this.hoistTemplateProperties(config.profiles, impConfig.baseProfile.type);
        }

        return { ...config, autoStore: true };
    }

    /**
     * Returns empty value that is appropriate for the property type.
     * @param propType The type of profile property
     * @returns Null or empty object
     */
    private static getDefaultValue(propType: string | string[]): any {
        // TODO How to handle profile property with multiple types
        if (Array.isArray(propType)) {
            propType = propType[0];
        }
        switch (propType) {
            case "string":  return "";
            case "number":  return 0;
            case "object":  return {};
            case "array":   return [];
            case "boolean": return false;
            default:        return null;
        }
    }

    /**
     * Moves duplicate default properties across child profiles into a base
     * profile.
     * @param profiles The profiles object of the config JSON
     * @param baseProfileType The type of the base profile
     */
    private static hoistTemplateProperties(profiles: { [key: string]: IConfigProfile }, baseProfileType: string): void {
        // Flatten properties into object that maps property name to list of values
        const flattenedProps: { [key: string]: any[] } = {};
        for (const childProfile of Object.values(profiles).filter(({ type }) => type !== baseProfileType)) {
            for (const [k, v] of Object.entries(childProfile.properties)) {
                flattenedProps[k] = [...(flattenedProps[k] || []), v];
            }
        }
        // List property names defined multiple times with the same value
        const duplicateProps: string[] = [];
        for (const [k, v] of Object.entries(flattenedProps)) {
            if (v.length > 1 && (new Set(v)).size === 1) {
                duplicateProps.push(k);
            }
        }
        // Remove duplicate properties from child profiles and store them in root profile
        for (const propName of duplicateProps) {
            profiles[baseProfileType].properties = {
                [propName]: flattenedProps[propName][0],
                ...profiles[baseProfileType].properties
            };
            for (const childProfile of Object.values(profiles).filter(({ type }) => type !== baseProfileType)) {
                delete childProfile.properties[propName];
            }
        }
    }
}
