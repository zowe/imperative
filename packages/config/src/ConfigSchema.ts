/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { ICommandProfileProperty } from "../../cmd";
import { IProfileProperty, IProfileSchema, IProfileTypeConfiguration } from "../../profiles";
import { IConfigSchema } from "./doc/IConfigSchema";

export class ConfigSchema {
    /**
     * JSON schema URI stored in $schema property of the schema
     * @readonly
     * @memberof ConfigSchema
     */
    private static readonly JSON_SCHEMA = "https://json-schema.org/draft/2020-12/schema";

    /**
     * Version number stored in $version property of the schema
     * @readonly
     * @memberof ConfigSchema
     */
    private static readonly SCHEMA_VERSION = 3;

    /**
     * Transform an Imperative profile schema to a JSON schema. Removes any
     * non-JSON-schema properties and translates anything useful.
     * @param schema The Imperative profile schema
     * @returns JSON schema for profile properties
     */
    private static generateJsonSchema(schema: IProfileSchema): any {
        const properties: { [key: string]: any } = {};
        const secureProps: string[] = [];
        for (const [k, v] of Object.entries(schema.properties || {})) {
            properties[k] = { type: v.type };
            const cmdProp = v as ICommandProfileProperty;
            if (cmdProp.optionDefinition != null) {
                properties[k].description = cmdProp.optionDefinition.description;
                if (cmdProp.optionDefinition.defaultValue != null) {
                    properties[k].default = cmdProp.optionDefinition.defaultValue;
                }
                if (cmdProp.optionDefinition.allowableValues != null) {
                    properties[k].enum = cmdProp.optionDefinition.allowableValues.values;
                }
            }
            if (v.secure) {
                secureProps.push(k);
            }
        }

        const propertiesSchema: any = {
            type: schema.type,
            title: schema.title,
            description: schema.description,
            properties
        };
        if (schema.required) {
            propertiesSchema.required = schema.required;
        }

        const secureSchema: any = {
            prefixItems: {
                enum: secureProps
            }
        };

        if (secureProps.length > 0) {
            return { properties: propertiesSchema, secure: secureSchema };
        } else {
            return { properties: propertiesSchema };
        }
    }

    /**
     * Transform a JSON schema to an Imperative profile schema.
     * @param schema The JSON schema for profile properties
     * @returns Imperative profile schema
     */
    private static parseJsonSchema(schema: any): IProfileSchema {
        const properties: { [key: string]: IProfileProperty } = {};
        for (const [k, v] of Object.entries((schema.properties.properties || {}) as { [key: string]: any })) {
            properties[k] = { type: v.type };
            if (schema.secure?.prefixItems.enum.includes(k)) {
                properties[k].secure = true;
            }
            if (v.description != null || v.default != null || v.enum != null) {
                (properties[k] as ICommandProfileProperty).optionDefinition = {
                    name: k,
                    type: v.type,
                    description: v.description,
                    defaultValue: v.default,
                    allowableValues: v.enum ? { values: v.enum } : undefined
                };
            }
        }

        return {
            title: schema.properties.title,
            description: schema.properties.description,
            type: schema.properties.type,
            properties,
            required: schema.properties.required
        };
    }

    /**
     * Dynamically builds the config schema for this CLI.
     * @param profiles The profiles supported by this CLI
     * @returns JSON schema for all supported profile types
     */
    public static buildSchema(profiles: IProfileTypeConfiguration[]): IConfigSchema {
        const entries: any[] = [];
        const defaultProperties: { [key: string]: any } = {};
        profiles.forEach((profile: { type: string, schema: IProfileSchema }) => {
            entries.push({
                if: { properties: { type: { const: profile.type } } },
                then: { properties: this.generateJsonSchema(profile.schema) }
            });
            defaultProperties[profile.type] = { type: "string" };
        });
        return {
            $schema: ConfigSchema.JSON_SCHEMA,
            $version: ConfigSchema.SCHEMA_VERSION,
            type: "object",
            description: "config",
            properties: {
                profiles: {
                    type: "object",
                    description: "named profiles config",
                    patternProperties: {
                        "^\\S*$": {
                            type: "object",
                            description: "a profile",
                            properties: {
                                type: {
                                    description: "the profile type",
                                    type: "string"
                                },
                                properties: {
                                    description: "the profile properties",
                                    type: "object"
                                },
                                profiles: {
                                    description: "additional sub-profiles",
                                    type: "object",
                                    $ref: "#/properties/profiles"
                                },
                                secure: {
                                    description: "secure property names",
                                    type: "array",
                                    prefixItems: {
                                        type: "string"
                                    },
                                    uniqueItems: true
                                }
                            },
                            allOf: entries
                        }
                    }
                },
                defaults: {
                    type: "object",
                    description: "default profiles config",
                    properties: defaultProperties
                }
            }
        };
    }

    /**
     * Loads Imperative profile schema objects from a schema JSON file.
     * @param schemaJson The schema JSON for config
     */
    public static loadProfileSchemas(schemaJson: IConfigSchema): IProfileTypeConfiguration[] {
        const patternName = Object.keys(schemaJson.properties.profiles.patternProperties)[0];
        const profileSchemas: IProfileTypeConfiguration[] = [];
        for (const obj of schemaJson.properties.profiles.patternProperties[patternName].allOf) {
            profileSchemas.push({
                type: obj.if.properties.type.const,
                schema: this.parseJsonSchema(obj.then.properties)
            });
        }
        return profileSchemas;
    }
}
