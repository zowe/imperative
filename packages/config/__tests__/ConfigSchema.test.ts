/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { IProfileTypeConfiguration } from "../../profiles";
import { ConfigSchema } from "../src/ConfigSchema";
import { cloneDeep } from "lodash";
import { ICommandProfileTypeConfiguration } from "../..";

describe("Config Schema", () => {
    const schema = ConfigSchema;
    const testProfileConfiguration: IProfileTypeConfiguration[] = [
        {
            type: "zosmf",
            schema: {
                title: "zosmf",
                description: "A fake zosmf profile",
                type: "object",
                required: [],
                properties: {
                    host: {
                        type: "string",
                        secure: false
                    }
                }
            }
        },
        {
            type: "base",
            schema: {
                title: "base",
                description: "A fake base profile",
                type: "object",
                required: [],
                properties: {
                    port: {
                        type: "number",
                        secure: false
                    }
                }
            }
        }
    ];
    const testProfileConfigurationSecure: IProfileTypeConfiguration[] = [
        {
            type: "zosmf",
            schema: {
                title: "zosmf",
                description: "A fake zosmf profile",
                type: "object",
                required: [],
                properties: {
                    host: {
                        type: "string",
                        secure: true
                    }
                }
            }
        }
    ];
    const testProfileConfigurationOptionDefinition: ICommandProfileTypeConfiguration[] = [
        {
            type: "zosmf",
            schema: {
                title: "zosmf",
                description: "A fake zosmf profile",
                type: "object",
                required: [],
                properties: {
                    host: {
                        type: "string",
                        optionDefinition: {
                            name: "host",
                            aliases: ["h"],
                            description: "The fake host to connect to",
                            type: "string",
                            defaultValue: "fake",
                            allowableValues: {
                                values: ["fake", "real"]
                            }
                        }
                    }
                }
            }
        }
    ];

    it("should be able to successfully build with no profile type configuration", () => {
        const testConfig: IProfileTypeConfiguration[] = [];
        const returnedSchema = schema.buildSchema(testConfig);
        const expectedAllOf: any = [];
        expect(returnedSchema).toMatchSnapshot();
        expect(returnedSchema.properties.profiles.patternProperties["^\\S*$"].allOf).toEqual(expectedAllOf)
    });

    it("should be able to successfully build with a single profile type configuration", () => {
        const testConfig: IProfileTypeConfiguration[] = cloneDeep(testProfileConfiguration);
        testConfig.pop();
        const returnedSchema = schema.buildSchema(testConfig);
        const expectedAllOf: any[] = [{
            if: {
                properties: {
                    type: {
                        const: "zosmf"
                    }
                }
            },
            then: {
                properties: {
                    properties: {
                        description: "A fake zosmf profile",
                        properties: {
                            host: {
                                type: "string"
                            },
                        },
                        required: [],
                        title: "zosmf",
                        type: "object"
                    }
                }
            }
        }];
        expect(returnedSchema).toMatchSnapshot();
        expect(returnedSchema.properties.profiles.patternProperties["^\\S*$"].allOf).toEqual(expectedAllOf)
    });

    it("should be able to successfully build with two profile type configurations", () => {
        const testConfig: IProfileTypeConfiguration[] = cloneDeep(testProfileConfiguration);
        const returnedSchema = schema.buildSchema(testConfig);
        const expectedAllOf: any[] = [
            {
                if: {
                    properties: {
                        type: {
                            const: "zosmf"
                        }
                    }
                },
                then: {
                    properties: {
                        properties: {
                            description: "A fake zosmf profile",
                            properties: {
                                host: {
                                    type: "string"
                                },
                            },
                            required: [],
                            title: "zosmf",
                            type: "object"
                        }
                    }
                }
            },
            {
                if: {
                    properties: {
                        type: {
                            const: "base"
                        }
                    }
                },
                then: {
                    properties: {
                        properties: {
                            description: "A fake base profile",
                            properties: {
                                port: {
                                    type: "number"
                                },
                            },
                            required: [],
                            title: "base",
                            type: "object"
                        }
                    }
                }
            }
        ];
        expect(returnedSchema).toMatchSnapshot();
        expect(returnedSchema.properties.profiles.patternProperties["^\\S*$"].allOf).toEqual(expectedAllOf)
    });

    it("should be able to successfully build with a secure single profile type configuration", () => {
        const testConfig: IProfileTypeConfiguration[] = cloneDeep(testProfileConfigurationSecure);
        const returnedSchema = schema.buildSchema(testConfig);
        const expectedAllOf: any[] = [{
            if: {
                properties: {
                    type: {
                        const: "zosmf"
                    }
                }
            },
            then: {
                properties: {
                    properties: {
                        description: "A fake zosmf profile",
                        properties: {
                            host: {
                                type: "string"
                            },
                        },
                        required: [],
                        title: "zosmf",
                        type: "object"
                    },
                    secure: {
                        prefixItems: {
                            enum: [
                                "host"
                            ]
                        }
                    }
                }
            }
        }];
        expect(returnedSchema).toMatchSnapshot();
        expect(returnedSchema.properties.profiles.patternProperties["^\\S*$"].allOf).toEqual(expectedAllOf)
    });

    it("should be able to successfully build with a complex single profile type configuration", () => {
        const testConfig: IProfileTypeConfiguration[] = cloneDeep(testProfileConfigurationOptionDefinition);
        const returnedSchema = schema.buildSchema(testConfig);
        const expectedAllOf: any[] = [{
            if: {
                properties: {
                    type: {
                        const: "zosmf"
                    }
                }
            },
            then: {
                properties: {
                    properties: {
                        description: "A fake zosmf profile",
                        properties: {
                            host: {
                                type: "string",
                                default: "fake",
                                description: "The fake host to connect to",
                                enum: ["fake", "real"]
                            },
                        },
                        required: [],
                        title: "zosmf",
                        type: "object"
                    }
                }
            }
        }];
        expect(returnedSchema).toMatchSnapshot();
        expect(returnedSchema.properties.profiles.patternProperties["^\\S*$"].allOf).toEqual(expectedAllOf)
    });

    it("should be able to regenerate profile schemas from a schema object", () => {
        const testConfig: IProfileTypeConfiguration[] = cloneDeep(testProfileConfiguration);
        const returnedSchema = schema.buildSchema(testConfig);
        const origSchemas = schema.loadProfileSchemas(returnedSchema);
        expect(origSchemas.length).toBe(2);
        expect(origSchemas[0].type).toBe(testConfig[0].type);
        expect(origSchemas[1].type).toBe(testConfig[1].type);
        // The comparison below needs to be done in this order since we only want to check the structure of the object
        expect(testConfig[0].schema).toMatchObject(origSchemas[0].schema);
        expect(testConfig[1].schema).toMatchObject(origSchemas[1].schema);
    });

    it("should be able to regenerate profile schemas with option definitions from a schema object", () => {
        const testConfig: IProfileTypeConfiguration[] = cloneDeep(testProfileConfigurationOptionDefinition);
        const returnedSchema = schema.buildSchema(testConfig);
        const origSchemas = schema.loadProfileSchemas(returnedSchema);
        expect(origSchemas.length).toBe(1);
        expect(origSchemas[0].type).toBe(testConfig[0].type);
        // The comparison below needs to be done in this order since we only want to check the structure of the object
        expect(testConfig[0].schema).toMatchObject(origSchemas[0].schema);
    });
});
