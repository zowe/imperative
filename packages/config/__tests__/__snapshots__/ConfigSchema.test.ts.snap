// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Config Schema should be able to successfully build with a complex single profile type configuration 1`] = `
Object {
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$version": 3,
  "description": "config",
  "properties": Object {
    "defaults": Object {
      "description": "default profiles config",
      "properties": Object {
        "zosmf": Object {
          "type": "string",
        },
      },
      "type": "object",
    },
    "profiles": Object {
      "description": "named profiles config",
      "patternProperties": Object {
        "^\\\\S*$": Object {
          "allOf": Array [
            Object {
              "if": Object {
                "properties": Object {
                  "type": Object {
                    "const": "zosmf",
                  },
                },
              },
              "then": Object {
                "properties": Object {
                  "properties": Object {
                    "description": "A fake zosmf profile",
                    "properties": Object {
                      "host": Object {
                        "default": "fake",
                        "description": "The fake host to connect to",
                        "enum": Array [
                          "fake",
                          "real",
                        ],
                        "type": "string",
                      },
                    },
                    "required": Array [],
                    "title": "zosmf",
                    "type": "object",
                  },
                },
              },
            },
          ],
          "description": "a profile",
          "properties": Object {
            "profiles": Object {
              "$ref": "#/properties/profiles",
              "description": "additional sub-profiles",
              "type": "object",
            },
            "properties": Object {
              "description": "the profile properties",
              "type": "object",
            },
            "secure": Object {
              "description": "secure property names",
              "prefixItems": Object {
                "type": "string",
              },
              "type": "array",
              "uniqueItems": true,
            },
            "type": Object {
              "description": "the profile type",
              "type": "string",
            },
          },
          "type": "object",
        },
      },
      "type": "object",
    },
  },
  "type": "object",
}
`;

exports[`Config Schema should be able to successfully build with a secure single profile type configuration 1`] = `
Object {
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$version": 3,
  "description": "config",
  "properties": Object {
    "defaults": Object {
      "description": "default profiles config",
      "properties": Object {
        "zosmf": Object {
          "type": "string",
        },
      },
      "type": "object",
    },
    "profiles": Object {
      "description": "named profiles config",
      "patternProperties": Object {
        "^\\\\S*$": Object {
          "allOf": Array [
            Object {
              "if": Object {
                "properties": Object {
                  "type": Object {
                    "const": "zosmf",
                  },
                },
              },
              "then": Object {
                "properties": Object {
                  "properties": Object {
                    "description": "A fake zosmf profile",
                    "properties": Object {
                      "host": Object {
                        "type": "string",
                      },
                    },
                    "required": Array [],
                    "title": "zosmf",
                    "type": "object",
                  },
                  "secure": Object {
                    "prefixItems": Object {
                      "enum": Array [
                        "host",
                      ],
                    },
                  },
                },
              },
            },
          ],
          "description": "a profile",
          "properties": Object {
            "profiles": Object {
              "$ref": "#/properties/profiles",
              "description": "additional sub-profiles",
              "type": "object",
            },
            "properties": Object {
              "description": "the profile properties",
              "type": "object",
            },
            "secure": Object {
              "description": "secure property names",
              "prefixItems": Object {
                "type": "string",
              },
              "type": "array",
              "uniqueItems": true,
            },
            "type": Object {
              "description": "the profile type",
              "type": "string",
            },
          },
          "type": "object",
        },
      },
      "type": "object",
    },
  },
  "type": "object",
}
`;

exports[`Config Schema should be able to successfully build with a single profile type configuration 1`] = `
Object {
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$version": 3,
  "description": "config",
  "properties": Object {
    "defaults": Object {
      "description": "default profiles config",
      "properties": Object {
        "zosmf": Object {
          "type": "string",
        },
      },
      "type": "object",
    },
    "profiles": Object {
      "description": "named profiles config",
      "patternProperties": Object {
        "^\\\\S*$": Object {
          "allOf": Array [
            Object {
              "if": Object {
                "properties": Object {
                  "type": Object {
                    "const": "zosmf",
                  },
                },
              },
              "then": Object {
                "properties": Object {
                  "properties": Object {
                    "description": "A fake zosmf profile",
                    "properties": Object {
                      "host": Object {
                        "type": "string",
                      },
                    },
                    "required": Array [],
                    "title": "zosmf",
                    "type": "object",
                  },
                },
              },
            },
          ],
          "description": "a profile",
          "properties": Object {
            "profiles": Object {
              "$ref": "#/properties/profiles",
              "description": "additional sub-profiles",
              "type": "object",
            },
            "properties": Object {
              "description": "the profile properties",
              "type": "object",
            },
            "secure": Object {
              "description": "secure property names",
              "prefixItems": Object {
                "type": "string",
              },
              "type": "array",
              "uniqueItems": true,
            },
            "type": Object {
              "description": "the profile type",
              "type": "string",
            },
          },
          "type": "object",
        },
      },
      "type": "object",
    },
  },
  "type": "object",
}
`;

exports[`Config Schema should be able to successfully build with no profile type configuration 1`] = `
Object {
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$version": 3,
  "description": "config",
  "properties": Object {
    "defaults": Object {
      "description": "default profiles config",
      "properties": Object {},
      "type": "object",
    },
    "profiles": Object {
      "description": "named profiles config",
      "patternProperties": Object {
        "^\\\\S*$": Object {
          "allOf": Array [],
          "description": "a profile",
          "properties": Object {
            "profiles": Object {
              "$ref": "#/properties/profiles",
              "description": "additional sub-profiles",
              "type": "object",
            },
            "properties": Object {
              "description": "the profile properties",
              "type": "object",
            },
            "secure": Object {
              "description": "secure property names",
              "prefixItems": Object {
                "type": "string",
              },
              "type": "array",
              "uniqueItems": true,
            },
            "type": Object {
              "description": "the profile type",
              "type": "string",
            },
          },
          "type": "object",
        },
      },
      "type": "object",
    },
  },
  "type": "object",
}
`;

exports[`Config Schema should be able to successfully build with two profile type configurations 1`] = `
Object {
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$version": 3,
  "description": "config",
  "properties": Object {
    "defaults": Object {
      "description": "default profiles config",
      "properties": Object {
        "base": Object {
          "type": "string",
        },
        "zosmf": Object {
          "type": "string",
        },
      },
      "type": "object",
    },
    "profiles": Object {
      "description": "named profiles config",
      "patternProperties": Object {
        "^\\\\S*$": Object {
          "allOf": Array [
            Object {
              "if": Object {
                "properties": Object {
                  "type": Object {
                    "const": "zosmf",
                  },
                },
              },
              "then": Object {
                "properties": Object {
                  "properties": Object {
                    "description": "A fake zosmf profile",
                    "properties": Object {
                      "host": Object {
                        "type": "string",
                      },
                    },
                    "required": Array [],
                    "title": "zosmf",
                    "type": "object",
                  },
                },
              },
            },
            Object {
              "if": Object {
                "properties": Object {
                  "type": Object {
                    "const": "base",
                  },
                },
              },
              "then": Object {
                "properties": Object {
                  "properties": Object {
                    "description": "A fake base profile",
                    "properties": Object {
                      "port": Object {
                        "type": "number",
                      },
                    },
                    "required": Array [],
                    "title": "base",
                    "type": "object",
                  },
                },
              },
            },
          ],
          "description": "a profile",
          "properties": Object {
            "profiles": Object {
              "$ref": "#/properties/profiles",
              "description": "additional sub-profiles",
              "type": "object",
            },
            "properties": Object {
              "description": "the profile properties",
              "type": "object",
            },
            "secure": Object {
              "description": "secure property names",
              "prefixItems": Object {
                "type": "string",
              },
              "type": "array",
              "uniqueItems": true,
            },
            "type": Object {
              "description": "the profile type",
              "type": "string",
            },
          },
          "type": "object",
        },
      },
      "type": "object",
    },
  },
  "type": "object",
}
`;
